File: src/my_project/medium_problems/from101to150/binary_tree_zigzag_bfs.py
from typing import List, Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
        
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []
        res, temp, stack, flag=[], [], [root], 1
        while stack:
            for i in range(len(stack)):
                
                node=stack.pop(0)
                temp+=[node.val]
                print(temp)
                if node.left: stack+=[node.left]
                if node.right: stack+=[node.right]
            res+=[temp[::flag]]
            temp=[]
            flag*=-1
        return res
        
solution = Solution()

print(solution.zigzagLevelOrder(TreeNode(3,TreeNode(9),TreeNode(20,TreeNode(15),TreeNode(7)))))

File: src/my_project_typescript/medium_problems_typescript/from1to50/binaryTreeZigZagBFS.ts


 class TreeNode {
      val: number
      left: TreeNode | null
      right: TreeNode | null
      constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
     }
  }
 


function zigzagLevelOrder(root: TreeNode | null): number[][] {
    if (!root) return [];
    let stack: TreeNode[] = [root];
    const res: number[][] = [];
    let i = 0;
  
    while (stack.length) {
      const newStack: TreeNode[] = [];
      res.push([]);
  
      for (const node of stack) {
        res[i].push(node.val);
        if (node.left) newStack.push(node.left);
        if (node.right) newStack.push(node.right);
      }
  
      stack = newStack;
      i++;
    }
  
    for (let i = 1; i < res.length; i += 2) res[i].reverse();
    return res;
  }